support/cas-server-support-saml-idp-web/src/main/java/org/apereo/cas/support/saml/web/idp/profile/builders/enc/SamlIdPObjectSigner.java

```diff
@@ -381,17 +383,43 @@ private static AbstractCredential finalizeSigningCredential(final MutableCredent
     /**
      * Gets signing private key.
      *
+     * @param service - the service being requested
      * @return the signing private key
      * @throws Exception the exception
      */
-    protected PrivateKey getSigningPrivateKey() throws Exception {
+    protected PrivateKey getSigningPrivateKey(final SamlRegisteredService service) throws Exception {
         val samlIdp = casProperties.getAuthn().getSamlIdp();
-        val signingKey = samlIdPMetadataLocator.getSigningKey();
+        val signingKey = createServicePrivateKeyResource(service.getIssuerSigningKeyLocation());
         val privateKeyFactoryBean = new PrivateKeyFactoryBean();
         privateKeyFactoryBean.setLocation(signingKey);
-        privateKeyFactoryBean.setAlgorithm(samlIdp.getMetadata().getPrivateKeyAlgName());
+        if (StringUtils.isBlank(service.getIssuerSigningKeyLocation())) {
+            privateKeyFactoryBean.setAlgorithm(samlIdp.getMetadata().getPrivateKeyAlgName());
+        } else {
+            privateKeyFactoryBean.setAlgorithm(service.getIssuerSigningAlgorithm());
+        }
         privateKeyFactoryBean.setSingleton(false);
         LOGGER.debug("Locating signature signing key from [{}]", signingKey);
         return privateKeyFactoryBean.getObject();
     }
+
+    /**
+     * Create the private key.
+     *
+     * @param location - resource location of private key
+     * @return - Resource
+     * @throws Exception if key creation ran into an error
+     */
+    protected Resource createServicePrivateKeyResource(final String location) throws Exception {
+        if (StringUtils.isBlank(location)) {
+            return samlIdPMetadataLocator.getSigningKey();
+        }
+
+        if (location.startsWith(ResourceUtils.CLASSPATH_URL_PREFIX)) {
+            return new ClassPathResource(StringUtils.removeStart(location, ResourceUtils.CLASSPATH_URL_PREFIX));
+        }
+        if (location.startsWith(ResourceUtils.FILE_URL_PREFIX)) {
+            return new FileSystemResource(StringUtils.removeStart(location, ResourceUtils.FILE_URL_PREFIX));
+        }
+        return new FileSystemResource(location);
```

Comment:
This is a good start, but it needs a few more changes. SAML2 idp metadata and/or keys should always be fetched from a locator instance. It would be contradictory to store metadata artifacts in a database while requiring a signing key for a service to be available on disk. We'll need to modify locator instances to allow for this behavior at the API level first, and then sort of make them service-aware...and this change needs to be applied to all locator types across the codebase.

Reply:
For now, my suggestion would be to simply back out the signing-key changes until the API work is done and keep service signing-algorithm changes.